### PR TITLE
refactor: Truth graph metrics for Exa.TrkX with separate algorithm

### PR DESCRIPTION
**Pull request by benjaminhuth**
Refactors code for truth graph building into separate algorithm, also improve reader/writer infrastructure for graphs.

### CODE REVIEW

File: Examples/Algorithms/TrackFindingExaTrkX/src/TruthGraphBuilder.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include <Acts/Definitions/Units.hpp>
+#include <ActsExamples/TrackFindingExaTrkX/TruthGraphBuilder.hpp>
+
+using namespace Acts::UnitLiterals;
+
+namespace ActsExamples {
+
+ActsExamples::TruthGraphBuilder::TruthGraphBuilder(Config config,
```
Comment (andiwand) [inline Examples/Algorithms/TrackFindingExaTrkX/src/TruthGraphBuilder.cpp:16]:
```suggestion
TruthGraphBuilder::TruthGraphBuilder(Config config,
```

File: Examples/Algorithms/TrackFindingExaTrkX/src/TruthGraphBuilder.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+  m_inputSimhits.maybeInitialize(m_cfg.inputSimHits);
+  m_inputMeasSimhitMap.maybeInitialize(m_cfg.inputMeasurementSimHitsMap);
+
+  bool a = m_inputMeasParticlesMap.isInitialized();
+  bool b =
+      m_inputSimhits.isInitialized() && m_inputMeasSimhitMap.isInitialized();
+
+  if (!(a || b)) {
+    throw std::invalid_argument("Missing inputs, cannot build truth graph");
+  }
```
Comment (andiwand) [inline Examples/Algorithms/TrackFindingExaTrkX/src/TruthGraphBuilder.cpp:34]: should we also issue an exception if both is present?

File: Examples/Algorithms/TrackFindingExaTrkX/src/TruthGraphBuilder.cpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        truthGraph.push_back(track[i].spacePointIndex);
+        truthGraph.push_back(track[i + 1].spacePointIndex);
+      }
+    }
+  }
+
+  return truthGraph;
+}
+
+ActsExamples::ProcessCode ActsExamples::TruthGraphBuilder::execute(
```
Comment (andiwand) [inline Examples/Algorithms/TrackFindingExaTrkX/src/TruthGraphBuilder.cpp:173]:
```suggestion
ProcessCode TruthGraphBuilder::execute(
```

File: Examples/Framework/include/ActsExamples/EventData/Graph.hpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <cstdint>
+#include <vector>
+
+namespace ActsExamples {
+
+/// Lightweight graph representation for GNN debugging
+///
+struct Graph {
+  /// The edges-vector contains flattened edge-pairs.Usually, the indices
```
Comment (andiwand) [inline Examples/Framework/include/ActsExamples/EventData/Graph.hpp:19]:
```suggestion
  /// The edges-vector contains flattened edge-pairs. Usually, the indices
```

**Review by andiwand — APPROVED** [review]
👍 left a few comments

File: Examples/Algorithms/TrackFindingExaTrkX/src/TruthGraphBuilder.cpp (hunk 2 above)
Comment (benjaminhuth) [reply Examples/Algorithms/TrackFindingExaTrkX/src/TruthGraphBuilder.cpp]: Changed to XOR operation

**Review by andiwand — APPROVED** [review]

**Review by andiwand — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :red_circle: Athena integration test results [24130951bf4ef8fc2412bd770c87734d41336385]
### Build job with this PR failed!
Please investigate the [build job](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/41091898) for the [pipeline](https://gitlab.cern.ch/acts/acts-athena-ci/-/pipelines/7734321)!

**Pull request merged**